src/lib/yacc/mod.rs

```diff
@@ -0,0 +1,439 @@
+    /// How many nonterminals does this grammar have?
+    nonterms_len: usize,
+    /// A mapping from NTIdx -> String.
+    nonterminal_names: Vec<String>,
+    /// A mapping from TIdx -> String.
+    terminal_names: Vec<String>,
+    /// A mapping from TIdx -> Option<Precedence>
+    terminal_precs: Vec<Option<Precedence>>,
+    /// How many terminals does this grammar have?
+    terms_len: usize,
+    /// The offset of the EOF terminal.
+    end_term: TIdx,
+    /// How many productions does this grammar have?
+    prods_len: usize,
+    /// Which production is the sole production of the start rule?
+    start_prod: PIdx,
+    /// A list of all productions.
+    prods: Vec<Vec<Symbol>>,
+    /// A mapping from rules to their productions. Note that 1) the order of rules is identical to
+    /// that of `nonterminal_names' 2) every rule will have at least 1 production 3) productions
+    /// are not necessarily stored sequentially.
+    rules_prods: Vec<Vec<PIdx>>,
+    /// A mapping from productions to their corresponding rule indexes.
+    prods_rules: Vec<NTIdx>,
+    /// The precedence of each production.
+    prod_precs: Vec<Option<Precedence>>
+}
+
+impl YaccGrammar {
+    /// Translate a `GrammarAST` into a `YaccGrammar`. This function is akin to the part a traditional
+    /// compiler that takes in an AST and converts it into a binary.
+    ///
+    /// As we're compiling the `GrammarAST` into a `Grammar` we do two extra things:
+    ///   1) Add a new start rule (which we'll refer to as "^", though the actual name is a fresh name
+    ///      that is guaranteed to be unique) that references the user defined start rule.
+    ///   2) Add a new end terminal (which we'll refer to as "$", though the actual name is a fresh
+    ///      name that is guaranteed to be unique).
+    /// So, if the user's start rule is S, we add a nonterminal with a single production:
+    ///   ^ : S '$';
+    pub fn new(ast: &ast::GrammarAST) -> YaccGrammar {
+        // The user is expected to have called validate before calling this function.
+        debug_assert!(ast.validate().is_ok());
+
+        // First of all generate guaranteed unique start nonterm and end term names. We simply keep
+        // making the string longer until we've hit something unique (at the very worst, this will
+        // require looping for as many times as there are nonterminals / terminals).
+
+        let mut start_nonterm = START_NONTERM.to_string();
+        while ast.rules.get(&start_nonterm).is_some() {
+            start_nonterm = start_nonterm + START_NONTERM;
+        }
+
+        let mut end_term = END_TERM.to_string();
+        while ast.tokens.iter().any(|x| x == &end_term) {
+            end_term = end_term + END_TERM;
+        }
+
+        let mut nonterm_names: Vec<String> = Vec::with_capacity(ast.rules.len() + 1);
+        nonterm_names.push(start_nonterm.clone());
+        for k in ast.rules.keys() { nonterm_names.push(k.clone()); }
```

Comment:
Personally, I find single-line for loops inside a larger block of code quite tricky to read...